lib/puppet/daemon.rb

```diff
@@ -109,8 +112,8 @@ def set_signal_traps
     signals.update({:HUP => :restart, :USR1 => :reload, :USR2 => :reopen_logs }) unless Puppet.features.microsoft_windows?
     signals.each do |signal, method|
       Signal.trap(signal) do
-        Puppet.notice "Caught #{signal}; calling #{method}"
-        send(method)
+        Puppet.notice "Caught #{signal}; storing #{method}"
+        @signals << method
```

Comment:
This isn't thread safe and could lead to a signal being lost. The signal handler and the consumer threads need to synchronize access to the shared Array. In practice that won't happen with MRI ruby because of its GIL. But it could happen on other ruby implementations, e.g. JRuby.


In practice this shouldn't be an issue, because MRI ruby still has a GIL. But that is not true for JRuby.